core/hashstorage.go

```diff
@@ -0,0 +1,153 @@
+// Copyright 2019 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+// Package core implements the Ethereum consensus protocol.
+
+package core
+
+import (
+	"bytes"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/log"
+)
+
+// For EVM execution, we need around 256 items. We add a few more to allow reorgs.
+// For mainnet, a couple more would suffice, but a few more added for
+// testnets/private nets
+const HashStorageElems = 288
+const HashStorageDataSize = 32 * HashStorageElems
+
+// HashStorage implements a storage for chains of hashes, intended to be used for quick lookup of blockhashes.
+// Internally, it uses an array of 8192 bytes to store 256 hashes in a circular buffer.
+// It enforces that all hashes added have a contiguous parent-child relation, and supports rollbacks
+// It is not thread-safe.
+type HashStorage struct {
+
+	// The data holds 256 32-byte hashes. The hashes are sequential, but also a
+	// circular buffer.
+	// The `head` points to the position of the latest hash.
+	// The parent, if present, is located 32 bytes back.
+	// [.., .., ..., head-2 , head-1, head, oldest, ... ]
+	data [HashStorageDataSize]byte
```

Comment:
Was there a reason to not make `data [HashStorageElems]common.Hash`? It seems like it would be more readable to me (unless read/write byte performance is that important?).

Reply:
Not really. Doing that would make the code easier to read, definitely. 
Originally, I wanted to do `setBytes` directly on the evm `big.Int` with as few allocs as possible. Then I opted for a more generic interface, and returned `common.Hash` instead. At that point, I might have just changed the interals too.. 

A secondary reason would be if we have a more generic usecase and want to store millions of hashes -- at that point it might we worthwhile to replace the mapping with a memory-mapped file, and that might be easier if the internals already use a byte array. 

But I agree that as it is right now, it should probably be changed.